resultset.go

```diff
@@ -11,3 +13,89 @@ type ResultBucket struct {
 	Buckets    map[string]*ResultBucket `json:"buckets"`
 	sourceRows []map[string]Cell
 }
+
+// ResultTable represents a Resultset split into row / columns at a depth.
+type ResultTable struct {
+	Rows         [][]map[string]interface{} `json:"rows"`
+	RowTitles    [][]string                 `json:"row_titles"`
+	ColumnTitles [][]string                 `json:"column_titles"`
+}
+
+// Tabulate takes a Resultset and converts it to tabular data.
+func Tabulate(results *Resultset, depth int) (*ResultTable, error) {
+	// Create our table.
+	table := &ResultTable{
+		Rows:         [][]map[string]interface{}{},
+		RowTitles:    [][]string{},
+		ColumnTitles: [][]string{},
+	}
+
+	// And a lookup helper instance.
+	lookup := &resultLookup{
+		cells:        map[string]map[string]interface{}{},
+		rowLookup:    map[string]bool{},
+		columnLookup: map[string]bool{},
+	}
+
+	// Recursively build the lookup for each of the root result buckets.
+	for _, bucket := range results.Buckets {
+		buildLookup([]string{}, 1, depth, table, lookup, bucket)
+	}
+
+	// Now build up the cells for each of the row / column tuples.
+	for _, row := range table.RowTitles {
+		tableRow := []map[string]interface{}{}
+		for _, column := range table.ColumnTitles {
+			tableRow = append(tableRow, lookup.cells[strings.Join(row, lookupKeyDelimiter)+lookupKeyDelimiter+strings.Join(column, lookupKeyDelimiter)])
+		}
+		table.Rows = append(table.Rows, tableRow)
+	}
+	// And we're done 👌.
+	return table, nil
+}
+
+// resultLookup stores specific data as the result set is recursively iterated over.
+type resultLookup struct {
+	cells        map[string]map[string]interface{}
+	rowLookup    map[string]bool
+	columnLookup map[string]bool
+}
+
+// lookupKeyDelimiter is used to flatten a string array to a single key.
+const lookupKeyDelimiter = "🔑🗝😡🗝🔑"
```

Comment:
This may actually be a really badly performant way of splitting this, but I wanted a delimiter that had the least chance of collision with row / column string values as possible.